src/agents/graph.py

```diff
@@ -0,0 +1,342 @@
+import json
+from collections.abc import Hashable
+from typing import Any, AsyncIterator, Dict, Literal, Protocol  # noqa UP
+
+from langchain_core.messages import AIMessage, HumanMessage
+from langchain_core.output_parsers import PydanticOutputParser
+from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
+from langgraph.checkpoint.base import BaseCheckpointSaver
+from langgraph.constants import END, START
+from langgraph.graph import MessagesState, StateGraph
+from langgraph.graph.graph import CompiledGraph
+
+from agents.common.agent import Agent
+from agents.common.data import Message
+from agents.common.state import AgentState, Plan, SubTask, UserInput
+from agents.common.utils import EXIT, exit_node, should_exit
+from agents.k8s.agent import K8S_AGENT, KubernetesAgent
+from agents.kyma.agent import KYMA_AGENT, KymaAgent
+from agents.supervisor.agent import COMMON, FINALIZER, SUPERVISOR, SupervisorAgent
+from utils.langfuse import handler
+from utils.logging import get_logger
+from utils.models import Model
+
+logger = get_logger(__name__)
+
+PLANNER = "Planner"
+
+
+def should_continue(state: MessagesState) -> Literal["action", "__end__"]:
+    """Return the next node to execute."""
+    last_message = state["messages"][-1]
+    return "action" if last_message.tool_calls else "__end__"
+
+
+class CustomJSONEncoder(json.JSONEncoder):
+    """Custom JSON encoder."""
+
+    def default(self, obj):  # noqa D102
+        if isinstance(obj, AIMessage | HumanMessage | SubTask):
+            return obj.__dict__
+        return super().default(obj)
+
+
+class Graph(Protocol):
+    """Graph interface."""
+
+    def astream(self, conversation_id: int, message: Message) -> AsyncIterator[str]:
+        """Stream the output to the caller asynchronously."""
+        ...
+
+
+class KymaGraph:
+    """Kyma graph class. Represents all the workflow of the application."""
+
+    model: Model
+    memory: BaseCheckpointSaver
+    supervisor_agent: Agent
+    kyma_agent: Agent
+    k8s_agent: Agent
+    members: list[str] = []
+
+    parser = PydanticOutputParser(pydantic_object=Plan)  # type: ignore
+
+    def __init__(self, model: Model, memory: BaseCheckpointSaver):
+        self.model = model
+        self.memory = memory
+
+        self.kyma_agent = KymaAgent(model)
+        self.k8s_agent = KubernetesAgent(model)
+        self.supervisor_agent = SupervisorAgent(
+            model, members=[KYMA_AGENT, K8S_AGENT, COMMON]
+        )
+
+        self.members = [self.kyma_agent.name, self.k8s_agent.name, COMMON]
+        self.graph = self._build_graph()
+
+    def _plan(self, state: AgentState) -> dict[str, Any]:
```

Comment:
Why do we have Supervisor and Planner separately? Isn't it Supervisor is supposed to the planning. IMO, we should merge the Supervisor and planner so we have one entity which is really controller and deciding the flow?

Reply:
As we discussed in the architecture meeting, if you have them in a single node it tries to re-plan again if they a single node. To avoid for the beginning I've put it outside. 
In the next iteration we optimize our nodes. 
I'll document this in the issue.